주스 메이커 [STEP3] - Groot, Borysarang

Groot-94 opened the pull request:
@protocorn93 

🙋‍♂️ 안녕하세요 콘! STEP3 첫 PR 보냅니다!
이번에도 좋은 리뷰 부탁드려요 :)

## 📌고민한점 
### View간 데이터 전달 방법
- `JuiceMakerViewController와 FruitStoreViewController간 데이터를 전송하는 방법에 대해서 어떻게 구현할지 고민했었습니다`. 저희는 활동학습때 공부했던 NotificationCenter를 활용하여 데이터의 변경점을 전파하도록 했습니다.
- 이렇게 했을 때 `장점은 stepper로 증감 및 쥬스를 제조하여 수량이 차감되는 즉시 객체에 전파되어 추가로 변경해주는 함수가 필요없이 수정이 가능하다는 점이라고 생각합니다`. `다만 Notification을 응용해보려고 사용했기에 1:N의 이벤트를 발생시키는 Notification의 특성을 살리지 못했다고 생각했습니다`. 저희의 코드는 JuiceMakerViewController와 FruitStoreViewController간에 FruitsStock 딕셔너리 한개만을 주고 받기 때문에 이런경우 추가적인 뷰가 생겨 전파를 동시에 두곳에서 진행하거나, KVO방식으로 데이터 변경을 알리는게 좀 더 낫지 않을까 생각했습니다.

### Stepper와 label 동기화

- 이번에 Stepper 컴포넌트를 처음 사용해보았습니다. `Stepper가 내부에 value 프로퍼티가 있었기에 이 value가 +값인지 -값인지를 나타내는 프로퍼티로 알았고 Label에 value값을 그대로 더해주는 방식으로 구현했었습니다.` 하지만 의도한대로 동작하지 않아서 다시 문서를 읽어본 결과 Stepper도 내부에 자체value를 가지고 있음을 알게 되었습니다. 따라서 `stepper의 value와 fruitsStock의 내부 저장 값을 일치시켜주니 의도한 대로 코드가 실행되었었습니다.`

### 앱 개발 시 지원대상(버전)을 정하는 기준

- Auto Layout을 적용하면서 가장 낮은 버전인 iPhone 4S 기기에서 테스트해보니 stepper의 크기가 비정상적으로 보이는 문제가 있었습니다. 이 부분에 대해 서포터와 이야기를 나눈 결과로, 너무 이전 버전의 기기에 대해선 고려하지 않아도 된다는 피드백을 받았습니다. 이 과정에서 실제로 `현업에서는 어떤 기준을 가지고 지원대상(버전)을 정하고 관리하는지 궁금증이 생겼습니다!`

### Stepper로 데이터 변경시 로직

- Stepper로 재고수정 시 `Stepper가 눌리는 순간 재고 변경을 post` 해줄지, 아니면 `modal이 닫히는 순간에 재고 변경을 post` 할지 고민했습니다.
- 고민 1: Stepper가 눌리는 순간 재고 변경을 post 하는 방식
>만약 Stepper가 10000번 눌리면 10000 번을 모두 post 하기 때문에 비효율적인 동작이라고 판단했습니다.
- 고민 2: modal이 닫히는 순간에 재고 변경을 post 하는 방식
>재고가 변경되지 않아도 post를 하게 되고 FruitStore에서 기존과 같은 재고를 다시 덮어쓰는 무의미한 동작이라고 판단했습니다.
- 해결방법: `modal이 닫히는 순간에 재고 변경을 post 하되, FruitStore에서 기존 재고와 변경된 부분에 대해서만 재고를 수정하는 방식으로 구현`

---
# 📌UML  
[ClassDiagram]
![](https://i.imgur.com/iHjYEPq.png)

[Flowchart]
![](https://i.imgur.com/YggRejI.jpg)

Groot-94 committed:
refactor: Refactoring updateFruitsStockLabels function

- change to forEach from compactMap

yusw10 committed:
design: Update FruitStoreView NavigationBar

- move to right backButton
- add backgroundColor at navigation bar

Groot-94 committed:
design: Update FruitStoreView

- include elements to Stack View

yusw10 committed:
design: modify juiceMaking buttons

- modify juice making button's font style
- add function to adjust button alignment

Groot-94 committed:
design: Modify FruitStoreViewController

- modify steppers

yusw10 committed:
feat: Regist UIComponents

- add Stepper to FruitStoreViewController
- add UILabel to FruitStoreViewController
- connect IBAction about stepper
- initialize temporary fruitsStock dictionary

Groot-94 committed:
feat: Add updateFruitsStockLabels, modifyFruitStockLabel function

- use function of JuiceMakerViewController

yusw10 committed:
feat: add updateStepperValue function

- add stepperDefaultValue for stepper
- add updateStepperValue function
- add stepper accessibliyIdentifier

Groot-94 committed:
feat: Implement pressStepper function

yusw10 committed:
feat: Add observer for delivering fruit stock dict

- add NotificationName for delivering fruit stock dictionary
- add addObserverFruitsStockDelivered function to regists observer

Groot-94 committed:
feat: Add postFruitsStockDelivered function of JuiceMakerViewController

- add viewWillDisappear function of JuiceMakerViewController
- modify modal Presentation Style to fullscreen

yusw10 committed:
feat: Add Updating fruits Stock

- add addObserverUpdatedFruitsStock function for updating fruit stocks
- add NotificationName for updating fruitsStock dict
- add post function of NotificationCenter at FruitStoreViewController

Groot-94 committed:
refactor: Modify naming of NotificationName

- modify to addObserverFruitsStockDidModified from addObserverUpdatedFruitsStock

yusw10 committed:
refactor: Refactoring Notification of FruitStoreVC

- refactor addObserverFruitsStockDidModified function to filtering modified fruitsStock
- add postFruitsStcokDidModidied function for notificate modifying

Groot-94 committed:
style: Modify overall code style

- Separating notification functions using extensions
- Grouping swift file

yusw10 committed:
docs: Update README.md

- edit step3 and step2

yusw10 committed:
Merge branch 'step3' of https://github.com/Groot-94/ios-juice-maker into step3

Groot-94 committed:
refactor: Hiding of function, modify naming of property

yusw10 committed:
docs: Modify UML and README.md

- add flowchart
- add modified ClassDiagram
- modify README.md
- add ClassDiagram drawio file

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -118,3 +119,15 @@ final class JuiceMakerViewController: UIViewController {
         }
     }
 }
+
```
Xcode Mark Line을 활용해서 코드의 물리적인 구조를 보다 읽기 좋은 구조로 바꿔볼 수 있어요! 

https://www.avanderlee.com/xcode/xcode-mark-line-comment/

yusw10 replied:
몰랐던 기능인데 말씀주셔서 한번 정리해봤습니다!
한가지 궁금한게 Mark Line도 git commit message convention처럼 양식같은게 존재할까요?(`refactor: Refactor function` 처럼)
아니면 다른사람이 이해 할수만 있다면 유도리 있게 작성하면 되는건가요?

[Mark Line 추가](https://github.com/Groot-94/ios-juice-maker/commit/66c4f14584b35d48ae83e36d02402454dc6b4505)
[1차 수정](https://github.com/yagom-academy/ios-juice-maker/pull/240/commits/1810048f744b73ee1a8d23154dc5eccd77ce0d07)

protocorn93 replied:
개인 스스로 기준을 세워도 무방해요! 다만 회사나 팀에 속하게 된다면 그 안의 컨벤션을 따라야겠죠! 저는 보통 

```swift
// MARK: - Actions 

extension SomeViewController { 
  @IBAction private func someAction1() { 
    .. 
  }

  @IBAction private func someAction2() { 
     .. 
  }

  .. 
}

// MARK: - UITableViewDelegate 

extension SomeViewController: UITableViewDelegate { 
  ...
}
```

이렇게 유사한 로직을 모아두거나 델리게이트 메서드 구현부를 분리해놓는데 사용하곤해요. Xcode 에디터를 통해 보다 쉽게 원하는 곳으로 이동할 수 있기 때문이죠!

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/FruitStoreViewController.swift:
```diff
@@ -8,12 +8,89 @@
+    private var fruitsStock: [Fruit: Int] = [:]
+    private let stepperDefaultValue: Int = 0
+    
+    @IBOutlet weak var strawberryStockLabel: UILabel!
+    @IBOutlet weak var bananaStockLabel: UILabel!
+    @IBOutlet weak var pineappleStockLabel: UILabel!
+    @IBOutlet weak var kiwiStockLabel: UILabel!
+    @IBOutlet weak var mangoStockLabel: UILabel!
+    
+    @IBOutlet weak var strawberryStepper: UIStepper!
+    @IBOutlet weak var bananaStepper: UIStepper!
+    @IBOutlet weak var pineappleStepper: UIStepper!
+    @IBOutlet weak var kiwiStepper: UIStepper!
+    @IBOutlet weak var mangoStepper: UIStepper!
```
`IBOutlet`도 접근제어자 추가해주시면 좋을 것 같아요~!

yusw10 replied:
private 추가 완료했습니다..!

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/FruitStoreViewController.swift:
```diff
@@ -8,12 +8,89 @@
+    }
+    
+    private func updateFruitsStockLabels(_ stock: [Fruit:Int]?) {
+        stock?.forEach { (key: Fruit, value: Int) in
+            modifyFruitStockLabel(key.rawValue, value)
```
`$0`, `$1`를 사용해서 코드를 보다 간결하게 작성해보실 수 있을 것 같아요

yusw10 replied:
`$0` 활용해봐서 key와 value 접근해보도록 변경했습니다..!

[고차함수 활용시 $0 사용](https://github.com/yagom-academy/ios-juice-maker/pull/240/commits/41320f7d86ca5a84b2da21c7863a3b1c28c60861)

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/FruitStoreViewController.swift:
```diff
@@ -8,12 +8,89 @@
+    
+    private func modifyFruitStockLabel(_ fruit: String, _ stock: Int) {
+        let allStockLabels: [UILabel] = [strawberryStockLabel, bananaStockLabel, pineappleStockLabel, kiwiStockLabel, mangoStockLabel]
+        for uiLabel in allStockLabels.filter({ $0.accessibilityIdentifier == fruit }) {
+            uiLabel.text = String(stock)
+        }
```
고차함수를 이어서 사용해볼 수 있을 것 같아요. filter 후 forEach로 구현해보는건 어떨까요?

yusw10 replied:
이부분도 위의 코멘트와 함께 같이 수정해보았습니다. filter와 `$0`를 사용했습니다.

[for loop에서 foreach로 refactor](https://github.com/yagom-academy/ios-juice-maker/pull/240/commits/41320f7d86ca5a84b2da21c7863a3b1c28c60861)

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -26,9 +26,15 @@ final class JuiceMakerViewController: UIViewController {
     
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(true)
+        postFruitsStockDelivered(juiceMaker.requestCurrentStock())
```
충분히 여기서 post 해줄 수 있다고 생각해요. 사용자 액션에 따른 행위로 단순히 화면이 닫힐 때보단 닫기 버튼이나 완료 버튼을 눌렀을 때 post 해주는게 보다 코드의 맥락을 파악하는데 도움이 될 것 같아요

Groot-94 replied:
콘의 말대로 사용자의 화면이 닫히는 부분에서 값을 전달하도록 수정했습니다. 
하지만, 이 부분에서 질문이 하나 생겼습니다! JuiceMaker에서 FruitStore modal을 띄어주는 부분은 2가지의 방식으로 구현이 필요합니다
1. 재고 수정 버튼을 클릭
2. 음료 주문했을 때 재고가 부족 시 뜨는 재고 수정 창


이렇게 두 가지 방법으로 나뉘는데 코드의 맥락을 위해서 모달을 띄어주는 함수 2개에서 값을 넘겨주도록 하는 건 같은 코드를 두 번 사용하는 방법이 아닐까 생각이 들었습니다. `이렇게 같은 View 이동을 다른 방식으로 구분해줘야 하는 상황에선 값을 넘겨주는 시점을 정할 때 View Life Cycle을 이용해서 구현하는지 궁금합니다!`
- JuiceMakerViewController -> FuitStoreViewController 
> [1f300f1](https://github.com/yagom-academy/ios-juice-maker/pull/240/commits/1f300f143cc35cd72ec8e49f3fea47b0aa77fe2f)
 - FuitStoreViewController -> JuiceMakerViewController
 > [5ae5ac8](https://github.com/yagom-academy/ios-juice-maker/pull/240/commits/5ae5ac8636b4aa23ad2a32ca9b59a40821ab3fd3)

protocorn93 replied:
이 부분에 대해선 아래에 좀 더 자세하게 작성해둘게요!

protocorn93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -118,3 +116,30 @@ final class JuiceMakerViewController: UIViewController {
+
+extension JuiceMakerViewController {
+    private func adjustButtonTitleAlignment(){
+        strawberryAndBananaJuiceButton.titleLabel?.textAlignment = .center
+        mangoAndKiwiJuiceButton.titleLabel?.textAlignment = .center
+        strawberryJuiceButton.titleLabel?.textAlignment = .center
+        bananaJuiceButton.titleLabel?.textAlignment = .center
+        pineappleJuiceButton.titleLabel?.textAlignment = .center
+        kiwiJuiceButton.titleLabel?.textAlignment = .center
+        mangoJuiceButton.titleLabel?.textAlignment = .center
```
버튼을 배열로 묶고 forEach로 값을 할당해보는건 어떨까요?

Groot-94 replied:
수정했습니다!
[41320f7](https://github.com/yagom-academy/ios-juice-maker/pull/240/commits/41320f7d86ca5a84b2da21c7863a3b1c28c60861)

protocorn93 reviewed (changes requested):
화면간 데이터 전달을 구현해보셨군요!! 두 분 모두 수고하셨어요! 👏👏

## View간 데이터 전달 방법

NotificationCenter는 아시다시피 1:N을 위해 사용될 수 있어요. 물론 좋은 도구이지만 그만큼 어디서나 쉽게 post하고 observe할 수 있어요. 의도하지 않은 post로 다른 화면에 영향을 미칠 수 있어요. 그렇기 때문에 사용하는데 주의하셔야 해요. 

저는 이렇게 이전 화면으로 데이터를 전달해주어야 할 경우에는 closure 혹은 delegate로 구현하곤 해요. 1:1 관계일 수 있고 그렇기에 두 객체간의 관계가 명확하기 때문이에요. 이번에는 delegate를 통해 구현해보시는건 어떨까요? 

## 앱 개발 시 지원대상(버전)을 정하는 기준 

보통 사내에선 지표를 보고 판단하곤해요. 버전별 사용자의 비율을 보고 올릴지말지 결정해요! 또한 보통 -2 버전까지 지원하더라구요. 현재 공식 iOS 버전이 15라면 13까지 지원하는 형식이에요. 

## Stepper로 데이터 변경시 로직

이 부분에 대해선 코멘트 확인 부탁드릴게요!

Groot-94 committed:
refactor: Refactoring function

- modify updateFruitsStockLabels
- modify modifyFruitStockLabel
- modify adjustButtonTitleAlignment

yusw10 committed:
refactor: Refactoring function

- modify updateFruitsStockLabel's indentation in FruitStroeViewController
- modify modifyFruitStockLabel's indentation in FruitStroeViewController
- modify adjustButtonTitleAlignment in FruitStroeViewController
- modify updateFruitsStockLabel in JuiceMakerViewController
- modify modifyFruitStockLabel in JuiceMakerViewController

Groot-94 committed:
refactor: Hiding, add Mark Lines

yusw10 committed:
-refactor: Refactoring to use Delegate Pattern

- add delegate protocol
- add deleagate at FruitStoreViewController
- add delegate at JuiceMakerViewController

Groot-94 committed:
Revert "refactor: Hiding, add Mark Lines"

This reverts commit 66c4f14584b35d48ae83e36d02402454dc6b4505.

# Conflicts:
#	JuiceMaker/JuiceMaker/Controller/FruitStoreViewController.swift

Groot-94 committed:
refactor: Hiding, remove Delegate Pattern

yusw10 committed:
refactor:  add Mark Lines

- add MARK Line
- refactor FruitDelegate arguments

Groot-94 committed:
refactor: Refectoring View Exchange function

yusw10 committed:
-style: Modify JuiceMakerVIewController Autolayout

- delete navigationController
- add Constraint at title and present button

Groot-94 committed:
refactor: Refactoring to use Delegate Pattern

yusw10 committed:
style: Modify Indentation and MarkLine

- modify Mark Line name to View Exchange
- modify Mark Line name to UIComponents data setting
- Unify spcing between functions

Groot-94 commented:
@protocorn93 
안녕하세요 콘! 좋은 리뷰 감사합니다!
리뷰를 보고 수정해봤습니다

## 질문 

### View간 데이터 전달

콘의 리뷰: `저는 이렇게 이전 화면으로 데이터를 전달해주어야 할 경우에는 closure 혹은 delegate로 구현하곤 해요. 1:1 관계일 수 있고 그렇기에 두 객체간의 관계가 명확하기 때문이에요. 이번에는 delegate를 통해 구현해보시는건 어떨까요?`

**그러면 혹시 콘은 초기 화면에서 다른 화면으로 값을 넘겨주는 방식은 어떤 방법으로 구현하시는지 궁금합니다!**

### NotifiCation

**실제로 NotifiCation을 어떤 부분에서 사용하는지, NotifiCation 사용을 지양해야 하는지 가 궁금합니다!**

Groot-94 committed:
docs: Add Sequence Diagram, update Class Diagram

yusw10 committed:
-docs: Modify README.md

- add STEP3
- add sequence diagram
- modify class diagram
- modify function documents

Groot-94 committed:
docs: Modify README.md

- modify style

protocorn93 commented:
> 그러면 혹시 콘은 초기 화면에서 다른 화면으로 값을 넘겨주는 방식은 어떤 방법으로 구현하시는지 궁금합니다!

먼저 이에 대해서 이야기해보기전에 화면을 구성하는 방법 두 가지에 대해 이야기 해보고 싶어요. 이에 따라 데이터를 전달하는 방법이 달라지거든요. 
1. 스토리보드를 통한 UI 구성 
2. 코드를 통한 UI 구성

두 가지 모두 장단점은 있지만 최근 대부분의 기업에선 코드를 통한 UI 구성 방법을 선택해서 사용하고 있어요. 혹은 두 가지를 병행하거나요! 이 두 가지 방법의 장단점은 구글링해보시면 상당히 많은 블로그와 사이트에서 다루고 있는 것을 확인해보실 수 있어요. 이를 통해서도 충분히 학습해보실 수 있을거라 생각됩니다. 

위애서 언급했듯이 UI 구성을 하는 방법에 따라 데이터를 전달하는 방법도 달라집니다. 

먼저 흔한 상황을 예로 들면 하나의 화면에서 다른 화면으로 초기값을 전달하고 그 화면에서 값이 변경되면 다시 이전 화면으로 돌려주는 상황이 있을수 있죠. A -> B, B -> A인 상황입니다. 여기서 중요한건 A는 B 화면을 알 수 있지만 B는 A 화면을 모르기 때문에 데이터를 전달하는 두 가지 상황에 있어 방법이 달라지는거죠. 

### A -> B

코드를 통해 UI를 그리면 기본적인 생성자를 통해 값을 쉽게 전달해줄 수 있어요. 이 역시 스토리보드 없이 UI 구성하는 방법에 대해 학습해보시는걸 권장드려요 

스토리보드를 사용한다면 가장 간단한 방법으로 Segue, property injection 등을 통해 값을 전달해줄 수 있어요. 

하지만 제가 학습을 권장드리는 방법은 iOS13부터 사용할 수 있게 된 [instantiateInitialViewController(creator:)](https://developer.apple.com/documentation/uikit/uistoryboard/3213988-instantiateinitialviewcontroller) 이 메서드를 사용해 스토리보드로 만든 뷰컨트롤러를 생성하는거에요. 사용법에 대해선 물론 학습을 해보셔야겠죠! 

또한 추가로 하나의 스토리보드에는 하나의 뷰컨트롤러만 위치시키는 것을 권장해드려요. 이를 통해 관리가 좀 더 쉬워지고 협업을 하는데 발생할 수 있는 충돌을 어느정도 방지할 수 있거든요. 


### B -> A

위에서 언급했듯이, B에서 이전 화면인 A로 전달하는 방법은 조금 달라요.B는 A를 모르기 때문이에요. 이를 위해 delegate이나 closure 그리고 Notification 등이 사용될 수 있는것이구요. 각각에 대해선 어느정도 학습이 되셨을거라 생각됩니다.

Notification 같은 경우에는 앱 전역으로 이벤트를 전달할 때 사용할 수 있겠죠! 당장 생각나는 예라면 금융앱이나 핀테크 앱에서 금액의 변화가 앱 전역에 영향을 미치는 경우가 되겠죠! 

지양하라는 것은 아니에요. 상황을 생각해보고 그에 맞게 사용하라는 말씀을 드리고 싶었어요! 현재 상황에서 Notification은 그렇게 좋은 선택지는 아닌 것 같아요!
